File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> res = new ArrayList<>();
        int n = nums.length;
        for(int i = 0 ; i < n ; i++){
          if(res.isEmpty() || res.getLast() < nums[i] ){
            res.addLast(nums[i]);
          } else{
            int l = 0;
            int r = res.size() - 1;
            
            while(l < r){
              int m = l + (r-l)/2;

              if(res.get(m) >= nums[i]){
                r = m;
              } else{
                l = m + 1;
              }
            }

            res.set(r , nums[i]);
          }
        }

        return res.size();
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)

        dp = [1 for _ in range(n)]

        for i in range(n):
            for j in range(i):
                if nums[j] < nums[i] and dp[i] < 1 + dp[j]:
                    dp[i] = 1 + dp[j]
        
        return max(dp)
        